moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h

```diff
@@ -124,8 +124,14 @@ class ActionBasedControllerHandle : public ActionBasedControllerHandleBase
 
   bool waitForExecution(const ros::Duration& timeout = ros::Duration(0)) override
   {
+    ros::Time end_time = ros::Time::now() + timeout;
     if (controller_action_client_ && !done_)
-      return controller_action_client_->waitForResult(timeout);
+    {
+      bool result = controller_action_client_->waitForResult(timeout);
+      while (ros::ok() && !done_ && (timeout.isZero() || ros::Time::now() < end_time))
+        ros::Duration(0.0001).sleep();  // wait for done callback (race-condition)
+      return result && (timeout.isZero() || ros::Time::now() < end_time);
```

Comment:
What's the reasoning behind the logic here? why include timeout.isZero()?

Reply:
If timeout isn't set, the thread should wait forever.